cs/src/Connections/WebSocketStream.cs

```diff
@@ -6,519 +6,547 @@
-            if (socket.State != WebSocketState.Open)
-            {
-                Close();
-            }
+            Close();
         }
+    }
 
-        /// <summary>
-        /// Current web socket close status or null if not closed.
-        /// </summary>
-        public WebSocketCloseStatus? CloseStatus
-        {
-            get => this.socket.CloseStatus;
-            set => this.closeStatus = value;
-        }
+    /// <summary>
+    /// Current web socket close status or null if not closed.
+    /// </summary>
+    public WebSocketCloseStatus? CloseStatus
+    {
+        get => this.socket.CloseStatus;
+        set => this.closeStatus = value;
+    }
 
-        /// <summary>
-        /// Current web socket close status description..
-        /// </summary>
-        public string? CloseStatusDescription
+    /// <summary>
+    /// Current web socket close status description..
+    /// </summary>
+    public string? CloseStatusDescription
+    {
+        get => this.socket.CloseStatusDescription;
+        set => this.closeStatusDescription = value;
+    }
+
+    /// <summary>
+    /// Connect to web socket.
+    /// </summary>
+    public static async Task<WebSocketStream> ConnectToWebSocketAsync(
+        Uri uri,
+        Action<ClientWebSocketOptions>? configure = default,
+        CancellationToken cancellation = default)
+    {
+        var socket = new ClientWebSocket();
+        try
         {
-            get => this.socket.CloseStatusDescription;
-            set => this.closeStatusDescription = value;
+            configure?.Invoke(socket.Options);
+            await socket.ConnectAsync(uri, cancellation);
+            return new WebSocketStream(socket);
         }
-
-        /// <summary>
-        /// Connect to web socket.
-        /// </summary>
-        public static async Task<WebSocketStream> ConnectToWebSocketAsync(Uri uri, Action<ClientWebSocketOptions>? configure = default, CancellationToken cancellation = default)
+        catch (WebSocketException wse)
+        when (wse.WebSocketErrorCode == WebSocketError.NotAWebSocket)
         {
-            var socket = new ClientWebSocket();
-            try
+            // ClientWebSocket.HttpStatusCode will be available in the future:
+            // https://github.com/dotnet/runtime/issues/25918
+            // Until then, the status code can be parsed from the exception message.
+            var statusCodeMessage = "server returned status code '";
+            var messageIndex = wse.Message.IndexOf(statusCodeMessage);
+            if (messageIndex > 0)
             {
-                configure?.Invoke(socket.Options);
-                await socket.ConnectAsync(uri, cancellation);
-                return new WebSocketStream(socket);
-            }
-            catch
-            {
-                socket.Dispose();
-                throw;
+                var offset = messageIndex + statusCodeMessage.Length;
```

Comment:
There is already code in `WebSocketStream` that gets the status code from the exception message, checked in with my PR https://github.com/microsoft/dev-tunnels/pull/99

Reply:
Yes, I just realized this is mostly duplicating that. I'll close this PR.